src/devices/GoPiGo3/GoPiGo3.cs

```diff
@@ -0,0 +1,887 @@
+                    GoPiGo3Info.Board = GetBoard();
+                    GoPiGo3Info.SoftwareVersion = GetFirmwareVersion();
+                    GoPiGo3Info.HardwareVersion = GetHardwareVersion();
+                    GoPiGo3Info.Id = GetIdHex();
+                }
+                catch (IOException ex)
+                {
+                    throw new IOException($"No SPI response. GoPiGo3 with address {SpiAddress} not connected.", ex);
+                }
+            if ((GoPiGo3Info.Manufacturer != ManufacturerName) || (GoPiGo3Info.Board != BoardName))
+                throw new IOException($"GoPiGo3 with address {SpiAddress} not connected.");
+            MotorGearRatio = DefaultMotorGearRatio;
+            EncoderTicksPerRotation = DefaultEncoderTicksPerRotation;
+            // Initialise the 2 Groove sensors
+            GrooveSensor = new GrooveSensor[2]
+            {
+                new GrooveSensor(GroovePort.Groove1),
+                new GrooveSensor(GroovePort.Groove2)
+            };
+        }
+
+        /// <summary>
+        /// Reset everything
+        /// </summary>
+        public void Dispose()
+        {
+            // Reset groove sensors
+            SetGrooveType(GroovePort.Both, GrooveSensorType.Custom);
+            SetGrooveMode(GroovePort.Both, GrooveInputOutput.InputDigital);
+            // Turn off the motors
+            SetMotorPower(MotorPort.Both, (byte)MotorSpeed.Float);
+            // Reset the motor limits
+            SetMotorLimits(MotorPort.Both, 0, 0);
+            // Turn off the servos
+            SetServo(ServoPort.Both, 0);
+            // Turn off the LEDs
+            SetLed((byte)GoPiGo3Led.LedEyeLeft + (byte)GoPiGo3Led.LedEyeRight + (byte)GoPiGo3Led.LedBlinkerLeft + (byte)GoPiGo3Led.LedBlinkerRight, Color.Black);
+        }
+
+        #endregion
+
+        #region SPI transfer
+
+        /// <summary>
+        /// Conduct a SPI transaction
+        /// </summary>
+        /// <param name="dataOut">A byte array to send.The length of the array will determine how many bytes are transferred.</param>
+        /// <returns>Returns an array of the bytes read.</returns>
+        public byte[] SpiTransferArray(byte[] dataOut)
+        {
+            byte[] result = new byte[dataOut.Length];
+            _spiDevice.TransferFullDuplex(dataOut, result);
+            return result;
+        }
+
+        /// <summary>
+        /// Read a 32 bit value over SPI
+        /// </summary>
+        /// <param name="MessageType">The SPI message type</param>
+        /// <returns>Returns the value read</returns>
+        public int SpiRead32(SpiMessageType MessageType)
+        {
+            int retVal = -1;
+            byte[] outArray = { SpiAddress, (byte)MessageType, 0, 0, 0, 0, 0, 0 };
+            byte[] reply = SpiTransferArray(outArray);
+
+            if (reply[3] == CorrectData)
+            {
+                retVal = (int)(reply[4] << 24) | (reply[5] << 16) | (reply[6] << 8) | reply[7];
```

Comment:
You shouldn't be doing these convertions yourself because you need to take into account endianness. You can instead use [BinaryPrimitives](https://github.com/dotnet/corefx/blob/master/src/System.Memory/ref/System.Memory.cs#L266-L349) methods to make conversions from bytes read into int, short, long that will take into account the endianness.

Reply:
I though we had something for that but couldn't find it :smile:

Reply:
Same feedback applies to all the upcoming conversions that you are making, you should be using BinaryPrimitives instead.

Reply:
> I though we had something for that but couldn't find it

It is on the System.Memory contract, so that means that the project will need to add a reference to `<PackageReference Include="System.Memory" Version="4.5.2" />`

Reply:
Good. I will adjust the code to use this for the byte arrays to int32 and int16 (using the BigEndians one). For the other way, from the int to the byte array, I will keep the code as it is. I see a bit less value there as it will introduce a Span of bytes, another function to call. Happy to debate about it.